### PR TITLE
[Web] Sets up event infrastructure for predict/suggest messages.

### PR DESCRIPTION
**Pull request by jahorton**
This small PR is designed to provide hooks for use in development of suggestion-selection UI toward our predictive text goals.  Nothing's actually providing input to the `ModelManager.predict` method yet.

### CODE REVIEW

**Comment by keyman-server** [comment]
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/147144:id/kvk/index.html)

File: web/source/text/prediction/modelManager.ts [hunk 1]
```diff
@@ -104,7 +117,39 @@ namespace com.keyman.text.prediction {
...
+      var promise = this.currentPromise = this.lmEngine.predict(transform, context);
+
+      // We've already invalidated any suggestions resulting from any previously-existing Promise -
+      // may as well officially invalidate them via event.
+      keyman.util.callEvent(ModelManager.EVENT_PREFIX + "invalidate", null);
+
+      let mm = this;
+      promise.then(function(suggestions: Suggestion[]) {
+        if(promise == mm.currentPromise) {
+          keyman.util.callEvent(ModelManager.EVENT_PREFIX + "suggestions_ready", suggestions);
```
Comment (darcywong00) [inline web/source/text/prediction/modelManager.ts]: The convention in kmwkeyboards.ts is lower-case, no underscores `suggestionsready`

Maybe l.144 becomes `invalidatesuggestions`

File: web/source/text/prediction/modelManager.ts [hunk 2]
```diff
@@ -63,16 +82,30 @@ namespace com.keyman.text.prediction {
...
           this.loadModel(model);
+          stateChange += 1;
         }
+
+        if(stateChange == 0) {
+          console.warn("Unexpected lack of model state change in ModelManager.onKeyboardChange!");
+          return;
+        }
+
+        let changeParam: ModelChangeEnum = stateChange == 1 ? 'loaded' : (stateChange == 2 ? 'unloaded' : 'changed');
```
Comment (darcywong00) [inline web/source/text/prediction/modelManager.ts]: I found it a little confusing why you declared a `ModelChangeEnum` but then tracking the state with arithmetic. Would this alternative do the same (from ll 90-107)?

```suggestion
      if(this.currentModel !== model) {
        let stateChange: ModelChangeEnum = 'changed';
        
        if(this.currentModel) {
          this.unloadModel();
          stateChange = 'unloaded';
        }
        
        if(model) {
          this.loadModel(model);
          stateChange = 'loaded';
        }
        
        if(stateChange == 'changed') {
          console.warn("...");
          return;
        }
        
       keyman.util.callEvent(ModelManager.EVENT_PREFIX + 'modelchange', stateChange);        
```

Comment (jahorton) [reply web/source/text/prediction/modelManager.ts]: This won't do the same; `'changed'` is for when _both_ unloading _and_ loading occur.  In your suggestion, it's for when _neither_ occur.

The fact that both can happen together is why I used arithmetic to calculate it internally, as it's easier to work with bit-flags for that.  Of course, those are less understandable outside the method, hence swapping to `string` values for the actual event parameter.

Comment (jahorton) [reply web/source/text/prediction/modelManager.ts]: So, maybe the following?

```
        let changeParam: ModelChangeEnum = stateChange == 1 ? 'loaded' : (stateChange == 2 ? 'unloaded' : /* == 3 */ 'changed');
```

Comment (jahorton) [reply web/source/text/prediction/modelManager.ts]: Perhaps instead:

```suggestion
        let changeParam: ModelChangeEnum = stateChange == 1 ? 'loaded' : (stateChange == 2 ? 'unloaded' : /* == 3 */ 'changed');
```

Comment (darcywong00) [reply web/source/text/prediction/modelManager.ts]: for clarity, maybe rename `ModelChangeEnum` to 
```
export type ModelStateEnum = 'invalid' | 'loaded' | 'unloaded' | 'changed';
```
and adjust the "state machine"?

Comment (jahorton) [reply web/source/text/prediction/modelManager.ts]: I feel like that'd indicate that `invalid` is a state that should be handled by the event's callbacks - something that isn't the case.  Is there a reason for wanting to handle `invalid` that I'm missing?

Comment (jahorton) [reply web/source/text/prediction/modelManager.ts]: To be clear, `'invalid'` _should_ be logically impossible - the `console.warn` is there to facilitate debugging if/when it breaks during development work.  I would consider it a bug if `'invalid'` **ever** occurred in production.

**Pull request merged**